Locks

alainjobart opened the pull request:
WIP. I want to see the integration tests pass.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/youtube/vitess/1722)
<!-- Reviewable:end -->

alainjobart commented:
@enisoc ready for review. Sorry about the long PR, but changing one thing has a big impact all over. I feel much better about all this code now, as locking is enforced everywhere it makes sense.

@michael-berlin FYI if you're interested.

alainjobart commented:
(I probably want to merge topotools into topo now, but I'll do that as a separate PR later. tests will not be easy, as they rely on zk fake topo, that sits in between topo and topotools)

enisoc commented:
Reviewed 41 of 42 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 19 unresolved discussions.

---

*[go/vt/tabletmanager/rpc_agent.go, line 27 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KI_4CodBixjVwT9U8Vh:-KI_4CodBixjVwT9U8Vi:1864604464) ([raw file](https://github.com/youtube/vitess/blob/0e0489ae2d546b25b3124ceec5927cd9cf3debc1/go/vt/tabletmanager/rpc_agent.go#L27)):*
> ```Go
> const (
> 	// TabletActionPing checks a tablet is alive
> 	TabletActionPing TabletAction = "Ping"
> ```

What do we gain from maintaining a separate list of all these string constants?

---

*[go/vt/topo/keyspace.go, line 171 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZwLn7BrmFez2bnd6T:-KIZwLn7BrmFez2bnd6U:1773106828) ([raw file](https://github.com/youtube/vitess/blob/0e0489ae2d546b25b3124ceec5927cd9cf3debc1/go/vt/topo/keyspace.go#L171)):*
> ```Go
> func (ts Server) UpdateKeyspace(ctx context.Context, ki *KeyspaceInfo) error {
> 	// make sure it is locked first
> 	if err := CheckKeyspaceLocked(ctx, ki.keyspace); err != nil {
> ```

This prevents version-based optimistic updates. Any reason for that?

---

*[go/vt/topo/locks.go, line 34 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZ07v4p8vi1j758kO2:-KIZ07v55CPtjWw6zvel:1137432964) ([raw file](https://github.com/youtube/vitess/blob/7fc33838884975b1a0f35e04de62acc58481769d/go/vt/topo/locks.go#L34)):*
> ```Go
> 	// LockTimeout is the command line flag that introduces a shorter
> 	// timeout for locking topology structures.
> 	LockTimeout = flag.Duration("lock_timeout", DefaultLockTimeout, "timeout for acquiring topology locks")
> ```

This flag used to be defined in `actionnode`, which was arguably ok because only specific binaries would be expected to use that package.

I think moving it to `topo` is broadcasting the flag too widely, and violates the principle that flags shouldn't be defined in "libraries". The `topo` package is needed just to get the public interfaces, for example; that shouldn't pollute the user's binary with flags.

---

*[go/vt/topo/locks.go, line 37 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZ1QDIj0yDTELAqKaX:-KIZ1QDJ319EwO5hCFSQ:-375354509) ([raw file](https://github.com/youtube/vitess/blob/7fc33838884975b1a0f35e04de62acc58481769d/go/vt/topo/locks.go#L37)):*
> ```Go
> )
> 
> // Action is the type for all Lock objects
> ```

I don't understand this comment. Isn't Lock the type for Lock objects?

---

*[go/vt/topo/locks.go, line 47 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZ3OT4V3tePfRQlLo9:-KIZ3OT4V3tePfRQlLoA:1621478233) ([raw file](https://github.com/youtube/vitess/blob/7fc33838884975b1a0f35e04de62acc58481769d/go/vt/topo/locks.go#L47)):*
> ```Go
> 
> 	// ShardActionReparent handles reparenting of the shard
> 	ShardActionReparent = Action("ReparentShard")
> ```

Why do we need to enumerate all possible reasons we might lock? Each of these is only really used in one place, so it seems like needless boilerplate. Why can't we just have `NewLock(comment string)`?

---

*[go/vt/topo/locks.go, line 176 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZqcuNl1PBtGCYemYY:-KIZqcuNl1PBtGCYemYZ:-870837321) ([raw file](https://github.com/youtube/vitess/blob/7fc33838884975b1a0f35e04de62acc58481769d/go/vt/topo/locks.go#L176)):*
> ```Go
> //   * 'vtctl SourceShardAdd' and 'vtctl SourceShardDelete' emergency operations
> // * keyspace-wide schema changes
> func (l *Lock) LockKeyspace(ctx context.Context, ts Server, keyspace string) (context.Context, func(context.Context, *error), error) {
> ```

In normal usage, it looks like you're just creating a Lock and then using it only once, to call LockKeyspace() or LockShard(). Why not just have LockKeyspace() be a static function (topo.LockKeyspace()) then?

---

*[go/vt/topo/locks.go, line 464 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZnK9dL5NicGIBgmmp:-KIZnK9dL5NicGIBgmmq:-1769711544) ([raw file](https://github.com/youtube/vitess/blob/7fc33838884975b1a0f35e04de62acc58481769d/go/vt/topo/locks.go#L464)):*
> ```Go
> // ReparentShardLock returns a Lock
> func ReparentShardLock(operation string, masterElectAlias *topodatapb.TabletAlias) *Lock {
> 	return NewLock(ShardActionReparent, &ReparentShardArgs{
> ```

Do we need this much detail in the lock contents? As long as locks auto-expire with TTL if the process dies, you should never need to look at the contents of the lock file in topology.

---

*[go/vt/topo/locks.go, line 476 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZnno21CwAX3U2M2Li:-KIZnno39aBfAf0CsuLl:-995889931) ([raw file](https://github.com/youtube/vitess/blob/7fc33838884975b1a0f35e04de62acc58481769d/go/vt/topo/locks.go#L476)):*
> ```Go
> 
> // CheckShardLock returns a Lock
> func CheckShardLock() *Lock {
> ```

This is an unfortunate name compared to `CheckShardLocked()`. But as mentioned above, I'd be in favor of eliminating all these action-specific lock functions.

---

*[go/vt/topo/locks.go, line 203 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZy2xlNgDvgQ3u0q3j:-KIZy2xnhrflucc4300W:-1578880091) ([raw file](https://github.com/youtube/vitess/blob/0e0489ae2d546b25b3124ceec5927cd9cf3debc1/go/vt/topo/locks.go#L203)):*
> ```Go
> 		actionNode: l,
> 	}
> 	return ctx, func(finalCtx context.Context, finalErr *error) {
> ```

Can you think of a use case for when the caller would give us a different `finalCtx` than the `ctx` we give it?

---

*[go/vt/topo/locks.go, line 212 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZxDMScwltqPp2VcFR:-KIZxDMScwltqPp2VcFS:-1991862179) ([raw file](https://github.com/youtube/vitess/blob/0e0489ae2d546b25b3124ceec5927cd9cf3debc1/go/vt/topo/locks.go#L212)):*
> ```Go
> 		}
> 
> 		*finalErr = l.unlockKeyspace(finalCtx, ts, keyspace, lockPath, *finalErr)
> ```

Here and above, you should only update `*finalErr` if it's nil. Otherwise you hide any errors that the calling function is trying to return.

Should maybe also check if `finalErr` itself is nil, in case the caller doesn't care about any unlock errors?

---

*[go/vt/topo/shard.go, line 184 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZq4yBQ7oDtNWn0Mxd:-KIZq4yBQ7oDtNWn0Mxe:-2124522660) ([raw file](https://github.com/youtube/vitess/blob/9b93e53741f1b2fba5e9c57896b49b55940efa29/go/vt/topo/shard.go#L184)):*
> ```Go
> 	defer span.Finish()
> 
> 	var version int64 = -1
> ```

-1 means unconditional update. Is there a reason you're removing that?

---

*[go/vt/topo/shard.go, line 238 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZrA9Q3ZWpksKTmsDF:-KIZrA9Q3ZWpksKTmsDG:-2072541268) ([raw file](https://github.com/youtube/vitess/blob/7fc33838884975b1a0f35e04de62acc58481769d/go/vt/topo/shard.go#L238)):*
> ```Go
> 	// Lock the keyspace, because we'll be looking at ServedTypes.
> 	lock := KeyspaceCreateShardLock()
> 	ctx, unlock, lockErr := lock.LockKeyspace(ctx, ts, keyspace)
> ```

The comment above says "this should be called while holding the keyspace lock" which conflicts with this.

---

*[go/vt/topo/shard.go, line 355 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZsFDkYrFV-ev098VM:-KIZsFDlVh09myLVK-ym:-517133181) ([raw file](https://github.com/youtube/vitess/blob/7fc33838884975b1a0f35e04de62acc58481769d/go/vt/topo/shard.go#L355)):*
> ```Go
> //   because it's not used in the same context (vertical vs horizontal sharding)
> func (si *ShardInfo) UpdateSourceBlacklistedTables(ctx context.Context, tabletType topodatapb.TabletType, cells []string, remove bool, tables []string) error {
> 	if err := CheckKeyspaceLocked(ctx, si.keyspace); err != nil {
> ```

This function only updates the in-memory Shard struct. It shouldn't need any distributed locks, right?

---

*[go/vt/topo/shard.go, line 402 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZw4XsEukE0N1bz_sL:-KIZw4XtbzqstAfNFVwC:1662569705) ([raw file](https://github.com/youtube/vitess/blob/7fc33838884975b1a0f35e04de62acc58481769d/go/vt/topo/shard.go#L402)):*
> ```Go
> //   because it's not used in the same context (vertical vs horizontal sharding)
> func (si *ShardInfo) UpdateDisableQueryService(ctx context.Context, tabletType topodatapb.TabletType, cells []string, disableQueryService bool) error {
> 	if err := CheckKeyspaceLocked(ctx, si.keyspace); err != nil {
> ```

Same here.

---

*[go/vt/topo/shard.go, line 176 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KI_7bCv0Dig8qQujpRj:-KI_7bCv0Dig8qQujpRk:431898575) ([raw file](https://github.com/youtube/vitess/blob/0e0489ae2d546b25b3124ceec5927cd9cf3debc1/go/vt/topo/shard.go#L176)):*
> ```Go
> 
> // UpdateShard masks ts.Impl.UpdateShard so nobody is tempted to use it.
> func (ts Server) UpdateShard(ctx context.Context, si *ShardInfo) error {
> ```

You can change the method signature while hiding a method from an embedded field:

``` go
func (ts Server) UpdateShard() {
  panic("...")
}
```

This way, the compiler will catch any accidental use (because who in their right mind would try to call `UpdateShard()` with no arguments).

---

*[go/vt/wrangler/keyspace.go, line 157 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZzzxdtKnYHwRQ_g1m:-KIZzzxdtKnYHwRQ_g1n:-2133045629) ([raw file](https://github.com/youtube/vitess/blob/0e0489ae2d546b25b3124ceec5927cd9cf3debc1/go/vt/wrangler/keyspace.go#L157)):*
> ```Go
> 		}
> 		for _, si := range refreshShards {
> 			rec.RecordError(wr.RefreshTablesByShard(ctx, si, servedType, cells))
> ```

Incidentally, this function has a typo in its name and comment: s/Tables/Tablets/g

---

*[go/vt/wrangler/keyspace.go, line 580 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KI_11QrWhG-z0wYwlnG:-KI_11QrWhG-z0wYwlnH:-821582687) ([raw file](https://github.com/youtube/vitess/blob/0e0489ae2d546b25b3124ceec5927cd9cf3debc1/go/vt/wrangler/keyspace.go#L580)):*
> ```Go
> 
> 	// execute the migration
> 	err = wr.migrateServedFrom(ctx, ki, si, servedType, cells, reverse, filteredReplicationWaitTime)
> ```

Should be called `migrateServedFromLocked`?

---

*[go/vt/wrangler/shard.go, line 91 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KI_2lPWZQ0hE3EE02o4:-KI_2lPWZQ0hE3EE02o5:-1059883392) ([raw file](https://github.com/youtube/vitess/blob/0e0489ae2d546b25b3124ceec5927cd9cf3debc1/go/vt/wrangler/shard.go#L91)):*
> ```Go
> 	}
> 
> 	// lock the keyspace
> ```

Can you add a comment for why this requires a keyspace lock. I'm still not sure.

---

*[go/vt/wrangler/shard.go, line 238 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KI_32_M8b9FQdz4Boi4:-KI_32_M8b9FQdz4Boi5:-1522717730) ([raw file](https://github.com/youtube/vitess/blob/0e0489ae2d546b25b3124ceec5927cd9cf3debc1/go/vt/wrangler/shard.go#L238)):*
> ```Go
> 	// lock the keyspace
> 	lock := topo.UpdateShardLock()
> 	ctx, unlock, lockErr := lock.LockKeyspace(ctx, wr.ts, keyspace)
> ```

In general, it seems weird to lock and unlock the keyspace for each per-shard operation. If there's some migration happening, then all these per-shard operations should require the keyspace lock to already be held before calling them. That way you don't race with other migrations, all shards are updated atomically, and there are fewer lock/unlock ops overall.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1722)*
<!-- Sent from Reviewable.io -->

alainjobart commented:
Review status: all files reviewed at latest revision, 19 unresolved discussions.

---

*[go/vt/tabletmanager/rpc_agent.go, line 27 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KI_4CodBixjVwT9U8Vh:-KIcGjtUivUg04Wfvrx4:1062434631) ([raw file](https://github.com/youtube/vitess/blob/0e0489ae2d546b25b3124ceec5927cd9cf3debc1/go/vt/tabletmanager/rpc_agent.go#L27)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> What do we gain from maintaining a separate list of all these string constants?
</details>
Not much, really. I want to refactor this entire thing, and remove these guys. I want to:
- remove the WrapActionLock... guys. Each action individually can decide if they need to lock, the logic should not be in the gRPC stub for instance. The code would move to the agent.
- add a generic HandlePanic instead, as we have in other RPC libraries.
- remove these string constants.

But this is a bigger change, I'll do it as part of a different PR.

---

*[go/vt/topo/keyspace.go, line 171 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZwLn7BrmFez2bnd6T:-KIcHJNJcwJfQ3FazcNN:-1081544541) ([raw file](https://github.com/youtube/vitess/blob/0e0489ae2d546b25b3124ceec5927cd9cf3debc1/go/vt/topo/keyspace.go#L171)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> This prevents version-based optimistic updates. Any reason for that?
</details>
So we should either do lock / get / update, or just updateKeyspaceFields, but not both. Currently, all the code assumes lock / get / update (it's only in 4 places or so that we do these updates).

All the places that do updates now also need to have the lock (for concurrency reasons, to lock out other actions, not for preventing concurrent access to the Keyspace structure). So sticking with lock / get / update works for now.

---

*[go/vt/topo/locks.go, line 34 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZ07v4p8vi1j758kO2:-KIcU4OtbI6N-ueFZVsE:-452775685) ([raw file](https://github.com/youtube/vitess/blob/7fc33838884975b1a0f35e04de62acc58481769d/go/vt/topo/locks.go#L34)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> This flag used to be defined in `actionnode`, which was arguably ok because only specific binaries would be expected to use that package.
> 
> I think moving it to `topo` is broadcasting the flag too widely, and violates the principle that flags shouldn't be defined in "libraries". The `topo` package is needed just to get the public interfaces, for example; that shouldn't pollute the user's binary with flags.
</details>
I don't have a better place for it, it needs to be visible to tabletmanager and wrangler... I could duplicate it in both these libraries, but then vtcombo (which links with both) wouldn't work... What do you think?

---

*[go/vt/topo/locks.go, line 37 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZ1QDIj0yDTELAqKaX:-KIcHzo65O7QS8wJ-cWW:-1546316501) ([raw file](https://github.com/youtube/vitess/blob/7fc33838884975b1a0f35e04de62acc58481769d/go/vt/topo/locks.go#L37)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> I don't understand this comment. Isn't Lock the type for Lock objects?
</details>
Will remove this (see next comment), this is just a bad search&replace on the comment :)

---

*[go/vt/topo/locks.go, line 47 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZ3OT4V3tePfRQlLo9:-KIcI4EDwpdKOTt6jutj:124104546) ([raw file](https://github.com/youtube/vitess/blob/7fc33838884975b1a0f35e04de62acc58481769d/go/vt/topo/locks.go#L47)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> Why do we need to enumerate all possible reasons we might lock? Each of these is only really used in one place, so it seems like needless boilerplate. Why can't we just have `NewLock(comment string)`?
</details>
Agreed, fixed. A lot of this stuff is just copying the old code, it wasn't obvious what extra should be fixed. Thank you for finding all these. :)

---

*[go/vt/topo/locks.go, line 176 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZqcuNl1PBtGCYemYY:-KIcJnEphDm9nhXHrO9i:-1588338071) ([raw file](https://github.com/youtube/vitess/blob/7fc33838884975b1a0f35e04de62acc58481769d/go/vt/topo/locks.go#L176)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> In normal usage, it looks like you're just creating a Lock and then using it only once, to call LockKeyspace() or LockShard(). Why not just have LockKeyspace() be a static function (topo.LockKeyspace()) then?
</details>
I thought about this, then the static Lock method would have two extra arguments, the 'action string' and 'args interface{}'. But then, we can't really use the methods that create the lock (and have the action string, and the extra struct with parameters)...

Then, actually, it can be a method of topo.Server, which makes a lot more sense.

---

*[go/vt/topo/locks.go, line 464 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZnK9dL5NicGIBgmmp:-KIcP7UaHo5PyQ3plV-z:1279465185) ([raw file](https://github.com/youtube/vitess/blob/7fc33838884975b1a0f35e04de62acc58481769d/go/vt/topo/locks.go#L464)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> Do we need this much detail in the lock contents? As long as locks auto-expire with TTL if the process dies, you should never need to look at the contents of the lock file in topology.
</details>
The ZK implementation creates a file with the lock comment in the topology, and if a lock cannot be acquired on time because of another lock, it shows you the contents of the other lock. I find it very useful in unit tests, to know who is holding the locks. It helped me at least once in this refactor, so I'm hesitant to take it out.

On the other hand, maybe if the LockKeyspace takes a string and ... as its last parameters, it can just Sprintf that and use just a single string to describe the action. Overall that might remove a lot of stuff.

---

*[go/vt/topo/locks.go, line 476 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZnno21CwAX3U2M2Li:-KIcPsinJj2ljzb9-V3e:-241067474) ([raw file](https://github.com/youtube/vitess/blob/7fc33838884975b1a0f35e04de62acc58481769d/go/vt/topo/locks.go#L476)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> This is an unfortunate name compared to `CheckShardLocked()`. But as mentioned above, I'd be in favor of eliminating all these action-specific lock functions.
</details>
Will eliminate yes.

---

*[go/vt/topo/locks.go, line 203 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZy2xlNgDvgQ3u0q3j:-KIcLJw0QH-Rd-lROqhh:-284170307) ([raw file](https://github.com/youtube/vitess/blob/0e0489ae2d546b25b3124ceec5927cd9cf3debc1/go/vt/topo/locks.go#L203)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> Can you think of a use case for when the caller would give us a different `finalCtx` than the `ctx` we give it?
</details>
Actually no, you're right. At first, I thought we needed a different context to unlock (so a timeout on the original one wouldn't matter), but we already detach the context in unlock, so it's unnecessary. And all use cases use the same anyway. So removing this.

---

*[go/vt/topo/locks.go, line 212 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZxDMScwltqPp2VcFR:-KIcN-sXP3QKANWu2121:1471373766) ([raw file](https://github.com/youtube/vitess/blob/0e0489ae2d546b25b3124ceec5927cd9cf3debc1/go/vt/topo/locks.go#L212)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> Here and above, you should only update `*finalErr` if it's nil. Otherwise you hide any errors that the calling function is trying to return.
> 
> Should maybe also check if `finalErr` itself is nil, in case the caller doesn't care about any unlock errors?
</details>
that check is already done by the unlockKeyspace function. the idea there is that the error returned by the user is higher priority than an unlock error, but if there is an unlock error, we just log the unlock error, and preserve the user error. if there is only an unlock error, we return it.

---

*[go/vt/topo/shard.go, line 184 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZq4yBQ7oDtNWn0Mxd:-KIcQ5TptSLyLlQ7Z8Qm:-1843009398) ([raw file](https://github.com/youtube/vitess/blob/9b93e53741f1b2fba5e9c57896b49b55940efa29/go/vt/topo/shard.go#L184)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> -1 means unconditional update. Is there a reason you're removing that?
</details>
we only use UpdateShardFields now, so we always have the real version. And we don't want to force updates any more, as a TER for instance can change the MasterAlias while we have locked the keyspace for a resharding op.

---

*[go/vt/topo/shard.go, line 238 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZrA9Q3ZWpksKTmsDF:-KIcRA9rrtL7N3m4iZp2:-1769986211) ([raw file](https://github.com/youtube/vitess/blob/7fc33838884975b1a0f35e04de62acc58481769d/go/vt/topo/shard.go#L238)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> The comment above says "this should be called while holding the keyspace lock" which conflicts with this.
</details>
The comment was 100% outdated. Fixed it.

---

*[go/vt/topo/shard.go, line 355 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZsFDkYrFV-ev098VM:-KIcRGH5PmYe01M5dkvU:158333620) ([raw file](https://github.com/youtube/vitess/blob/7fc33838884975b1a0f35e04de62acc58481769d/go/vt/topo/shard.go#L355)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> This function only updates the in-memory Shard struct. It shouldn't need any distributed locks, right?
</details>
I don't want any code to call this without the keyspace lock, this is the whole point of adding these checks. I caught one use case where we weren't locking the keyspace but calling this function...

---

*[go/vt/topo/shard.go, line 402 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZw4XsEukE0N1bz_sL:-KIcRQrT5ytgMLQ3r25Y:1082548332) ([raw file](https://github.com/youtube/vitess/blob/7fc33838884975b1a0f35e04de62acc58481769d/go/vt/topo/shard.go#L402)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> Same here.
</details>
Same reply, I want to enforce the keyspace being locked when calling this.

---

*[go/vt/topo/shard.go, line 176 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KI_7bCv0Dig8qQujpRj:-KIcQ4Od7nQH1ifMEauB:-499146009) ([raw file](https://github.com/youtube/vitess/blob/0e0489ae2d546b25b3124ceec5927cd9cf3debc1/go/vt/topo/shard.go#L176)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> You can change the method signature while hiding a method from an embedded field:
> 
> ``` go
> func (ts Server) UpdateShard() {
>   panic("...")
> }
> ```
> 
> This way, the compiler will catch any accidental use (because who in their right mind would try to call `UpdateShard()` with no arguments).
</details>
Done.

---

*[go/vt/wrangler/keyspace.go, line 157 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZzzxdtKnYHwRQ_g1m:-KIcRfJWPF6CktgCl3ZY:-499146009) ([raw file](https://github.com/youtube/vitess/blob/0e0489ae2d546b25b3124ceec5927cd9cf3debc1/go/vt/wrangler/keyspace.go#L157)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> Incidentally, this function has a typo in its name and comment: s/Tables/Tablets/g
</details>
Done.

---

*[go/vt/wrangler/keyspace.go, line 580 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KI_11QrWhG-z0wYwlnG:-KIcRsDjujyzqOQvLTZw:800479996) ([raw file](https://github.com/youtube/vitess/blob/0e0489ae2d546b25b3124ceec5927cd9cf3debc1/go/vt/wrangler/keyspace.go#L580)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> Should be called `migrateServedFromLocked`?
</details>
Why yes it should.

---

*[go/vt/wrangler/shard.go, line 91 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KI_2lPWZQ0hE3EE02o4:-KIcTCVGiTDy8JYPAQuF:1578509997) ([raw file](https://github.com/youtube/vitess/blob/0e0489ae2d546b25b3124ceec5927cd9cf3debc1/go/vt/wrangler/shard.go#L91)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> Can you add a comment for why this requires a keyspace lock. I'm still not sure.
</details>
Done. Also added comments on the Shard fields in the proto to be clearer.

---

*[go/vt/wrangler/shard.go, line 238 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KI_32_M8b9FQdz4Boi4:-KIcTNI21DwlWCu6tIXd:1454988075) ([raw file](https://github.com/youtube/vitess/blob/0e0489ae2d546b25b3124ceec5927cd9cf3debc1/go/vt/wrangler/shard.go#L238)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> In general, it seems weird to lock and unlock the keyspace for each per-shard operation. If there's some migration happening, then all these per-shard operations should require the keyspace lock to already be held before calling them. That way you don't race with other migrations, all shards are updated atomically, and there are fewer lock/unlock ops overall.
</details>
That's what we do, in the high level operations, like MigrateServedTypes(master, ...) we lock the keyspace, and do a bunch of stuff, including deleting SourceShard objects.

However, these specific methods that 'seem weird' are all emergency vtctl commands in case something goes horribly wrong with the higher level process. It's the case for SourceShardDelete/Add, SetServedTypes, ... They are all emergency manual operations that shouldn't be needed in normal operation, but we provide just in case.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1722)*
<!-- Sent from Reviewable.io -->

alainjobart commented:
OK I think I addressed all comments, with one leftover change for later to review the tabletamanager RPCWrap stuff and simplify that, and another to figure out where the flag should go.

@enisoc thanks a lot for the detailed review!

enisoc commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> except for some remaining concerns about `finalErr` handling, and some optional nits.

<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary>
> OK I think I addressed all comments, with one leftover change for later to review the tabletamanager RPCWrap stuff and simplify that, and another to figure out where the flag should go.
> 
> 
> 
> @enisoc thanks a lot for the detailed review!
> 
> 
</details>


---

Reviewed 9 of 9 files at r3, 9 of 9 files at r4.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[go/vt/topo/locks.go, line 34 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZ07v4p8vi1j758kO2:-KIdOFFl3YUQqw0_pjrG:-1678304430) ([raw file](https://github.com/youtube/vitess/blob/7fc33838884975b1a0f35e04de62acc58481769d/go/vt/topo/locks.go#L34)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary>
> I don't have a better place for it, it needs to be visible to tabletmanager and wrangler... I could duplicate it in both these libraries, but then vtcombo (which links with both) wouldn't work... What do you think?
</details>
I can't think of an easy solution that's better, so we can revisit later. It's just unfortunate that we're going farther from the ideal instead of getting closer or moving laterally.

It would be slightly better if we rename this to "topo_lock_timeout" so it's less likely to conflict if some other binary needs to import the topo package. But I'm not sure if it's going to cause more upgrade headaches than it's worth. Up to you.

---

*[go/vt/topo/locks.go, line 176 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZqcuNl1PBtGCYemYY:-KIdKhnB3S0lwXLiIdak:-754228000) ([raw file](https://github.com/youtube/vitess/blob/7fc33838884975b1a0f35e04de62acc58481769d/go/vt/topo/locks.go#L176)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary>
> I thought about this, then the static Lock method would have two extra arguments, the 'action string' and 'args interface{}'. But then, we can't really use the methods that create the lock (and have the action string, and the extra struct with parameters)...
> 
> Then, actually, it can be a method of topo.Server, which makes a lot more sense.
</details>
Ah of course it should be a method on topo.Server. Don't know why I didn't think of that. :)

---

*[go/vt/topo/locks.go, line 212 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZxDMScwltqPp2VcFR:-KIdLcPZIdsgBhE_iQzr:833982042) ([raw file](https://github.com/youtube/vitess/blob/0e0489ae2d546b25b3124ceec5927cd9cf3debc1/go/vt/topo/locks.go#L212)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary>
> that check is already done by the unlockKeyspace function. the idea there is that the error returned by the user is higher priority than an unlock error, but if there is an unlock error, we just log the unlock error, and preserve the user error. if there is only an unlock error, we return it.
</details>
Ah I didn't notice that `*finalErr` is passed as a parameter as well. It seems like it'd be cleaner though if you handle the masking/logging here, and let `unlockKeyspace()` return errors normally. Then you wouldn't need to check `actionError` so often, and it would be clearer that this guarantee is enforced.

---

*[go/vt/topo/locks.go, line 133 \[r4\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIdMqBZhYbhQD5V48PY:-KIdMqB_Awc6vNJM_zkL:1669897048) ([raw file](https://github.com/youtube/vitess/blob/7394e2a9fcae63df804333ef06a745a348877b50/go/vt/topo/locks.go#L133)):*
> ```Go
> 
> 	// lock
> 	l := newLock(fmt.Sprintf(format, v...))
> ```

I guess this is convenient, but it's kind of surprising that a function called `LockKeyspace()` would have the semantics of a format function, as if it were `LockKeyspacef()`. This can result in unexpected results if, for example, you're not trying to use the format feature but your string happens to have percent signs in it. I would prefer to just let each call site decide if they want to use `Sprintf` or not on their own, but I don't feel strongly about that if you like this better.

---

*[go/vt/topo/locks.go, line 149 \[r4\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIdNhkib0oTOefbt1e1:-KIdNhkib0oTOefbt1e2:1249070460) ([raw file](https://github.com/youtube/vitess/blob/7394e2a9fcae63df804333ef06a745a348877b50/go/vt/topo/locks.go#L149)):*
> ```Go
> 
> 		if _, ok := i.info[keyspace]; !ok {
> 			*finalErr = fmt.Errorf("trying to unlock keyspace %v multiple times", keyspace)
> ```

This one will hide the action error, which I don't think we want to do.

---

*[go/vt/topo/shard.go, line 355 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZsFDkYrFV-ev098VM:-KIdbZhwDvItZWxuKr31:-2068977971) ([raw file](https://github.com/youtube/vitess/blob/7fc33838884975b1a0f35e04de62acc58481769d/go/vt/topo/shard.go#L355)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary>
> I don't want any code to call this without the keyspace lock, this is the whole point of adding these checks. I caught one use case where we weren't locking the keyspace but calling this function...
</details>
I see. It just seemed weird to me that we're checking at the point where we update the in-memory struct, rather than checking when we interact with the topo service. But I see what you're saying, that after this the only chance we'd have to catch it is in UpdateShardFields and we can't just always require a keyspace lock there. So this is the only way.

It should be documented in the comment then that this function can only be called while holding the keyspace lock. Same for UpdateDisableQueryService() below.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1722#-:-KIdd7BnwAaK01CgVgA2:578298792)*
<!-- Sent from Reviewable.io -->

alainjobart commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[go/vt/topo/locks.go, line 34 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZ07v4p8vi1j758kO2:-KIdec825W1GmLM6jb_X:1767900370) ([raw file](https://github.com/youtube/vitess/blob/7fc33838884975b1a0f35e04de62acc58481769d/go/vt/topo/locks.go#L34)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> I can't think of an easy solution that's better, so we can revisit later. It's just unfortunate that we're going farther from the ideal instead of getting closer or moving laterally.
> 
> It would be slightly better if we rename this to "topo_lock_timeout" so it's less likely to conflict if some other binary needs to import the topo package. But I'm not sure if it's going to cause more upgrade headaches than it's worth. Up to you.
</details>
I don't mind renaming it, I think it's unspecified anywhere, we just use the default. But I'd rather do that in a separate change though.

---

*[go/vt/topo/locks.go, line 212 \[r2\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZxDMScwltqPp2VcFR:-KIdgXdel10va9L_1PyP:-499146009) ([raw file](https://github.com/youtube/vitess/blob/0e0489ae2d546b25b3124ceec5927cd9cf3debc1/go/vt/topo/locks.go#L212)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> Ah I didn't notice that `*finalErr` is passed as a parameter as well. It seems like it'd be cleaner though if you handle the masking/logging here, and let `unlockKeyspace()` return errors normally. Then you wouldn't need to check `actionError` so often, and it would be clearer that this guarantee is enforced.
</details>
Done.

---

*[go/vt/topo/locks.go, line 133 \[r4\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIdMqBZhYbhQD5V48PY:-KIdemckeBUvAA8ya4bY:-746837308) ([raw file](https://github.com/youtube/vitess/blob/7394e2a9fcae63df804333ef06a745a348877b50/go/vt/topo/locks.go#L133)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> I guess this is convenient, but it's kind of surprising that a function called `LockKeyspace()` would have the semantics of a format function, as if it were `LockKeyspacef()`. This can result in unexpected results if, for example, you're not trying to use the format feature but your string happens to have percent signs in it. I would prefer to just let each call site decide if they want to use `Sprintf` or not on their own, but I don't feel strongly about that if you like this better.
</details>
I was also thinking about this. I will make this change.

---

*[go/vt/topo/locks.go, line 149 \[r4\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIdNhkib0oTOefbt1e1:-KIdflQeCP9pKWrD0mTx:-499146009) ([raw file](https://github.com/youtube/vitess/blob/7394e2a9fcae63df804333ef06a745a348877b50/go/vt/topo/locks.go#L149)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> This one will hide the action error, which I don't think we want to do.
</details>
Done.

---

*[go/vt/topo/shard.go, line 355 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1722#-KIZsFDkYrFV-ev098VM:-KIdj1A3UN_AjKYLqGW5:-499146009) ([raw file](https://github.com/youtube/vitess/blob/7fc33838884975b1a0f35e04de62acc58481769d/go/vt/topo/shard.go#L355)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> I see. It just seemed weird to me that we're checking at the point where we update the in-memory struct, rather than checking when we interact with the topo service. But I see what you're saying, that after this the only chance we'd have to catch it is in UpdateShardFields and we can't just always require a keyspace lock there. So this is the only way.
> 
> It should be documented in the comment then that this function can only be called while holding the keyspace lock. Same for UpdateDisableQueryService() below.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1722)*
<!-- Sent from Reviewable.io -->

enisoc commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> :lgtm: except for some remaining concerns about `finalErr` handling, and some optional nits.
</details>


---

Reviewed 6 of 6 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1722#-:-KIdmb0ci8AtIq29cKkL:1417088721)*
<!-- Sent from Reviewable.io -->
